src/dist/manifest.rs

```diff
@@ -588,21 +589,37 @@ impl Component {
         }
     }
 
-    pub(crate) fn new_with_target(pkg_with_target: &str, is_extension: bool) -> Option<Self> {
-        for (pos, _) in pkg_with_target.match_indices('-') {
-            let pkg = &pkg_with_target[0..pos];
-            let target = &pkg_with_target[pos + 1..];
-            if let Some(partial) = PartialTargetTriple::new(target) {
-                if let Ok(triple) = TargetTriple::try_from(partial) {
-                    return Some(Self {
-                        pkg: pkg.to_string(),
-                        target: Some(triple),
-                        is_extension,
-                    });
-                }
+    pub(crate) fn try_new(
```

Comment:
IMO neither the PR description nor the commit message make it clear what this change is trying to achieve. I could go dig through the issue thread, but that's pretty inefficient. Can you make sure it's a little easier to address what these changes are designed to do? I see a substantial increase in complexity here and I think such a change should be well-motivated.

Reply:
@djc Are you suggesting that at the top of this PR there should be a short briefing/rationale part?

Reply:
> @djc Are you suggesting that at the top of this PR ([#3601 (comment)](https://github.com/rust-lang/rustup/pull/3601#issue-2055114196)) there should be a short briefing/rationale part? If so I agree with this point, but the code changes per se looks fine to me.

Yes, basically.

Reply:
@ongchi Could you please add a rationale section in https://github.com/rust-lang/rustup/pull/3601#issue-2055114196? Thanks a lot!